crates/planner/src/logical/bind.rs

```diff
@@ -0,0 +1,414 @@
+                || t.is_integer()
+                || t.is_float()
+                || t.is_string()
+                || t.is_bytes()
+                || t.is_identity()
+                || t.is_address() =>
+        {
+            Ok(expr)
+        }
+        (op, ty) => Err(ConstraintViolation::op(op, ty).into()),
+    }
+}
+
+fn assert_eq_types(a: &Type, b: &Type) -> TypingResult<()> {
+    if a == b {
+        Ok(())
+    } else {
+        Err(unexpected_type(a, b))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use spacetimedb_lib::{db::raw_def::v9::RawModuleDefV9Builder, AlgebraicType, ProductType};
+    use spacetimedb_primitives::TableId;
+    use spacetimedb_schema::{def::ModuleDef, schema::TableSchema};
+    use std::sync::Arc;
+
+    use super::{parse_and_type_sub, SchemaView};
+
+    fn module_def() -> ModuleDef {
+        let mut builder = RawModuleDefV9Builder::new();
+        builder.build_table_with_new_type(
+            "t",
+            ProductType::from([
+                ("u32", AlgebraicType::U32),
+                ("f32", AlgebraicType::F32),
+                ("str", AlgebraicType::String),
+                ("arr", AlgebraicType::array(AlgebraicType::String)),
+            ]),
+            true,
+        );
+        builder.build_table_with_new_type(
+            "s",
+            ProductType::from([
+                ("id", AlgebraicType::identity()),
+                ("u32", AlgebraicType::U32),
+                ("arr", AlgebraicType::array(AlgebraicType::String)),
+                ("bytes", AlgebraicType::bytes()),
+            ]),
+            true,
+        );
+        builder.finish().try_into().expect("failed to generate module def")
+    }
+
+    struct SchemaViewer(ModuleDef);
+
+    impl SchemaView for SchemaViewer {
+        fn schema(&self, name: &str, _: bool) -> Option<Arc<TableSchema>> {
+            self.0.table(name).map(|def| {
+                Arc::new(TableSchema::from_module_def(
+                    def,
+                    TableId(if *def.name == *"t" { 0 } else { 1 }),
+                ))
+            })
+        }
+    }
+
+    #[test]
+    fn valid() {
+        let tx = SchemaViewer(module_def());
+
+        for sql in [
+            "select * from t",
+            "select * from t where true",
+            "select * from t where t.u32 = 1",
+            "select * from t where t.u32 = 1 or t.str = ''",
+            "select * from s where s.bytes = 0xABCD",
+            "select * from s where s.bytes = X'ABCD'",
+            "select * from s as r where r.bytes = 0xABCD",
+            "select * from (select t.* from t join s)",
+            "select * from (select t.* from t join s on t.u32 = s.u32 where t.f32 = 0.1)",
+            "select * from (select t.* from t join (select s.u32 from s) s on t.u32 = s.u32)",
+        ] {
+            let result = parse_and_type_sub(sql, &tx).inspect_err(|_| {
+                // println!("sql: {}\n\n\terr: {}\n", sql, err);
+            });
```

Comment:
I think this way of testing could blindly let pass an unrelated error and mask something is broken.
 
If the idea is to make checking for the error easier without heavy matching, how about adding a simple enum `Error{TypeCheck, FieldNotFound}` that we return for the error?